vimeo/vimeo-networking-java
# Support enterprise SSO login

# Summary
We want to support login for enterprise customers via SSO. This is an internal use only API, but is very similar in usage to the publicly accessible code grant authentication API (via `createCodeGrantAuthorizationUri` and `authenticateWithCodeGrant`).

The API consists of 3 functions that were added to the `Authenticator`.
- `fetchSsoDomain(domain: String, VimeoCallback<SsoDomain>)`
- `createSsoGrantAuthorizationUri(domain: SsoDomain, responseCode: String): String`
- `ssoCodeGrant(uri: String, marketingOptIn: Boolean, VimeoCallback<VimeoAccount>)`

`createSsoGrantAuthorizationUri` is very similar to `createCodeGrantAuthorizationUri` in both its definition and implementation. The main difference is that the SSO version requires an `SsoDomain` instance. The reason for this is that while the base code grant auth URI is fixed for public code grants, for enterprise SSO the base URI is provided by the API. As a result, the consumer needs to make a request to `fetchSsoDomain` before creating the auth URI.

`ssoCodeGrant` is almost identical to `authenticateWithCodeGrant` in definition, implementation, and usage. The only difference in definition is that the marketing opt in is used. The implementation differs in that it hits a different endpoint. The usage is the same in that it requires the consumer to receive a callback with a redirect URI and then pass that redirect URI to the `ssoCodeGrant` function.

## Changes
- Added `SsoDomain` DTO.
- Added functions to `Authenticator` mentioned in above summary and counterparts in the `AuthService`.
- Added `url` property to a `VimeoCall` that allows the `HttpUrl` of the request to be returned.
- Added convenience `enqueueWithAccountStore` extension function on `VimeoCall<VimeoAccount>` that stores the result of a successful request as the currently logged in account.
- Extracted all string literals in `AuthService` into constants (not sure if this is necessarily a good idea since it makes seeing the actual values more difficult?)
- Added new errors to `AuthParam`

## Testing
NOTE: `:example` can be easily configured to test this.

1. Configure the client with an id and secret that has access to enterprise SSO login. 
2. Make sure to set the redirect URI in the configuration. 
3. Configure the app to handle the redirect URI via an `intent-filter` in the manifest.
4. Make a request to `fetchSsoDomain` with a domain that is supported by enterprise login.
5. Create the auth URL using `createSsoGrantAuthorizationUri` and the domain received back from the previous request.
6. Direct the user to log in via the browser using the URL created.
7. Receive the redirect URL back.
8. Log in using `ssoCodeGrant` and the URL that you received in the redirect.
